swap/swap.go

```diff
@@ -17,69 +17,340 @@
 	"strconv"
 	"sync"
+	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/contracts/swap"
+	cswap "github.com/ethersphere/swarm/contracts/swap"
 	"github.com/ethersphere/swarm/log"
 	"github.com/ethersphere/swarm/p2p/protocols"
 	"github.com/ethersphere/swarm/state"
 )
 
-// SwAP Swarm Accounting Protocol
+// ErrInvalidChequeSignature indicates the signature on the cheque was invalid
+var ErrInvalidChequeSignature = errors.New("invalid cheque signature")
+
+// Swap represents the SwAP Swarm Accounting Protocol
 // a peer to peer micropayment system
 // A node maintains an individual balance with every peer
 // Only messages which have a price will be accounted for
 type Swap struct {
-	stateStore state.Store        //stateStore is needed in order to keep balances across sessions
-	lock       sync.RWMutex       //lock the balances
-	balances   map[enode.ID]int64 //map of balances for each peer
+	api                 PublicAPI
+	stateStore          state.Store          // stateStore is needed in order to keep balances across sessions
+	lock                sync.RWMutex         // lock the balances
+	balances            map[enode.ID]int64   // map of balances for each peer
+	cheques             map[enode.ID]*Cheque // map of balances for each peer
+	peers               map[enode.ID]*Peer   // map of all swap Peers
+	backend             cswap.Backend        // the backend (blockchain) used
+	owner               *Owner               // contract access
+	params              *Params              // economic and operational parameters
+	contractReference   *swap.Swap           // reference to the smart contract
+	oracle              PriceOracle          // the oracle providing the ether price for honey
+	paymentThreshold    int64                // balance difference required for requesting cheque
+	disconnectThreshold int64                // balance difference required for dropping peer
+}
+
+// Owner encapsulates information related to accessing the contract
+type Owner struct {
+	Contract   common.Address    // address of swap contract
+	address    common.Address    // owner address
+	privateKey *ecdsa.PrivateKey // private key
+	publicKey  *ecdsa.PublicKey  // public key
+}
+
+// Params encapsulates param
+type Params struct {
+	InitialDepositAmount uint64 //
+}
+
+// NewDefaultParams returns a Params struct filled with default values
+func NewDefaultParams() *Params {
+	return &Params{
+		InitialDepositAmount: DefaultInitialDepositAmount,
+	}
 }
 
 // New - swap constructor
-func New(stateStore state.Store) (swap *Swap) {
-	swap = &Swap{
-		stateStore: stateStore,
-		balances:   make(map[enode.ID]int64),
+func New(stateStore state.Store, prvkey *ecdsa.PrivateKey, contract common.Address, backend cswap.Backend) *Swap {
+	sw := &Swap{
```

Comment:
check line 98

Reply:
Agree with @holisticode.

Reply:
disagree with both. `createOwner` should not be a receiver method on `Swap` since it does not use the `Swap` struct at all.
then you can just add the `owner: createOwner(prvKey, contract),` to the literal

Reply:
on closer look, i think @acud is right. @holisticode do you still object to this change?

Reply:
No, I agree to have that changed.

Reply:
fixed, thanks.